keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md

```diff
@@ -99,105 +101,167 @@ A developer has a Windows web app that they would like to deploy in a container
+
+  - Check the pod spec for inlined credspec JSON in the pod and per-container annotations [`pod.alpha.kubernetes.io/windows-gmsa-credspec` and `container.windows-gmsa-credspec.alpha.kubernetes.io/container-name` for Alpha] or securityContext fields [`podspec.securityContext.windows.gmsaCredSpec` and `podspec.container[i].securityContext.windows.gmsaCredSpec` Beta onwards]. These may exist as a result of the app admin optionally specifying the credspec JSON (besides the name).
+    - If a credspec JSON is not specified already, GMSAAuthorizer will look up the corresponding GMSACredSpec name and use the credspec JSON prepared from the previous step to populate the annotations or securityContext fields for credspec JSON.
+    - If a credspec JSON is specified, look up the corresponding GMSACredSpec name and perform a deep equal check of the members of the specified credspec JSON and prepared credspec JSON from previous step. If the deep equal check fails, GMSAAuthorizer will be fail pod creation with 400: BadRequest.
+
+  - Check the specified ServiceAccount is authorized for the `use` verb on all specified GMSACredSpecs at the pod level and for each container. If the authorization check fails due to absence of requisite RBAC roles, the pod creation will be failed with a 403: Forbidden. 
+
+7. The Windows CRI implementation will access the credspec JSON through annotations [`container.windows-gmsa-credspec.alpha.kubernetes.io/container-name` for Alpha] or securityContext fields [`podspec.container[i].securityContext.windows.gmsaCredSpec` Beta onwards] for each container in a pod and copy the contents to the [OCI windows.CredentialSpec](https://github.com/opencontainers/runtime-spec/blob/master/config-windows.md#credential-spec) field. Please see the Implementation section below for details on the enhancements necessary.
+
+8. The Windows OCI implementation validates the credspec and fails to start the container if its invalid or access to the GMSA from the node is denied. The containers that fail to start due to invalid GMSA or inaccessible AD will end up in a failed status with the following detail when described:
+```
+State:              Waiting
+      Reason:           CrashLoopBackOff
+Last State:         Terminated
+      Reason:           ContainerCannotRun
+      Message:          [runtime specific error e.g. "encountered an error during CreateContainer:..."]
```

Comment:
it is not clear to me that we will provide a very specific error message that GMSA is what failed to start the container. we def need to do that.

Reply:
What I meant above is that the specific error in processing the GMSA is detected and returned to Kubernetes CRI layer by the container runtime. The container runtime can fail to create a container due to several reasons like incorrect Windows base image, etc besides invalid GMSA config. I will clarify this point above.

I think the container runtime specific error message is slightly beyond the scope of the Kubernetes KEP but we should audit all run times that support GMSA is reporting appropriate error strings to orchestrators that hcsshim in Windows is reporting to them.

Reply:
I have added a note on how init containers maybe used to get around those scenarios where containers get started even though the credspec configuration is incorrect.